editor/modes/visual-editor/block.js

```diff
@@ -251,68 +251,161 @@ class VisualEditorBlock extends Component {
 
 	onKeyDown( event ) {
 		const { keyCode, target } = event;
+		const { previousBlock, nextBlock, onFocus } = this.props;
+		const up = keyCode === UP;
+		const down = keyCode === DOWN;
+		const left = keyCode === LEFT;
+		const right = keyCode === RIGHT;
+
+		const isVerticalEdge = ( { editor, reverse } ) => {
+			const rangeRect = editor.selection.getBoundingClientRect();
+			const buffer = rangeRect.height / 2;
```

Comment:
Rather than trying to place the caret at the very top or bottom (which might be off by 0-1px), it seems better to attempt to place it in the middle of the line.